ArchiSteamFarm/Bot.cs

```diff
@@ -2849,16 +2849,59 @@ public sealed class Bot : IAsyncDisposable {
 			}
 		}
 
-		private async Task<HashSet<uint>> GetCompletedBadgeAppIDs() {
-			// TODO - extract logic from CardsFarmer and check all pages containing less than **150** craftable badges
-			IDocument? badgePage = await ArchiWebHandler.GetBadgePage(1).ConfigureAwait(false);
+		private async Task<ISet<uint>> GetCompletedBadgeAppIDs() {
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(1).ConfigureAwait(false);
 
 			if (badgePage == null) {
 				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
 
 				return new HashSet<uint>(0);
 			}
 
+			ISet<uint> appIDs = GetCompletedBadgeAppIDs(badgePage);
+
+			byte maxPages = 1;
+			IElement? htmlNode = badgePage.SelectSingleNode("(//a[@class='pagelink'])[last()]");
+
+			if (htmlNode != null) {
+				string lastPage = htmlNode.TextContent;
+
+				if (string.IsNullOrEmpty(lastPage)) {
+					ArchiLogger.LogNullError(nameof(lastPage));
+
+					return new HashSet<uint>(0);
+				}
+
+				if (!byte.TryParse(lastPage, out maxPages) || (maxPages == 0)) {
+					ArchiLogger.LogNullError(nameof(maxPages));
+
+					return new HashSet<uint>(0);
+				}
+			}
+
+			const byte maxBadgesPerPage = 150;
+			byte currentPage = 2;
+
+			while ((currentPage < maxBadgesPerPage) && (appIDs.Count == currentPage * maxBadgesPerPage)) {
+				appIDs.UnionWith(await GetCompletedBadgeAppIDs(currentPage++).ConfigureAwait(false));
+			}
+
+			return appIDs;
+		}
+
+		private async Task<ISet<uint>> GetCompletedBadgeAppIDs(byte page) {
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(page).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return new HashSet<uint>(0);
+			}
+
+			return GetCompletedBadgeAppIDs(badgePage);
+		}
+
+		private ISet<uint> GetCompletedBadgeAppIDs(IDocument badgePage) {
```

Comment:
Don't return interface types. The rule should always be:
- Accept as arguments the most generic type, for example `IReadOnlyCollection<byte>` or even `IEnumerable<byte>` instead of `List<byte>`
- Always return the most specific type, for example `HashSet<uint` and not `ISet<uint>`.

The only exception from the second would be with public APIs if you don't know better what specific type the response will be of, I don't see that here, it's `HashSet<uint>`.